CoreScriptsRoot/Modules/VR/Panel3D.lua

```diff
@@ -0,0 +1,760 @@
+		lookAtGuiElement = nil,
+		isClosest = true
+	}
+
+	if panels[name] then
+		error("A panel by the name of " .. name .. " already exists.")
+	end
+	panels[name] = instance
+
+	return setmetatable(instance, Panel.mt)
+end
+
+--Panel accessor methods
+function Panel:GetPart()
+	if not self.part then
+		self.part = Util.Create "Part" {
+			Name = self.name,
+			Parent = nil,
+
+			Transparency = 1,
+
+			CanCollide = false,
+			Anchored = true,
+			Archivable = false,
+
+			Size = Vector3.new(1, 1, 1)
+		}
+	end
+	return self.part
+end
+
+function Panel:GetGUI()
+	if not self.gui then
+		local part = self:GetPart()
+		self.gui = Util.Create "SurfaceGui" {
+			Parent = CoreGui,
+			Adornee = part,
+			Active = true,
+			ToolPunchThroughDistance = 1000,
+			CanvasSize = Vector2.new(0, 0),
+			Enabled = self.isEnabled
+		}
+		--todo: remove this pcall when api is live
+		pcall(function()
+			self.gui.AlwaysOnTop = true
+		end)
+	end
+	return self.gui
+end
+
+function Panel:FindHoveredGuiElement(elements)
+	local x, y = self.lookAtPixel.X, self.lookAtPixel.Y
+	for i, v in pairs(elements) do
+		local minPt = v.AbsolutePosition
+		local maxPt = v.AbsolutePosition + v.AbsoluteSize
+		if minPt.X <= x and maxPt.X >= x and
+		   minPt.Y <= y and maxPt.Y >= y then
+			return v, i
+		end
+	end
+end
+--End of panel accessor methods
+
+
+--Panel update methods
+function Panel:SetPartCFrame(cframe)
+	if not self.part then
+		return
+	end
+	self.part.CFrame = cframe * CFrame.new(0, 0, -0.5)
+end
+
+function Panel:SetEnabled(enabled)
+	if self.isEnabled == enabled then
+		return
+	end
+
+	self.isEnabled = enabled
+	if not enabled and (not self.part or not self.gui) then
+		return
+	else
+		self:GetPart()
+		self:GetGUI()
+	end
+	if enabled then
+		self.part.Parent = workspace.CurrentCamera --todo: Perhaps this can change soon.
+		self.gui.Enabled = true
+		for i, v in pairs(self.subpanels) do
+			v.part.Parent = workspace.CurrentCamera
+			v.gui.Enabled = true
+		end
+	else
+		self.part.Parent = nil
+		self.gui.Enabled = false
+		for i, v in pairs(self.subpanels) do
+			v.part.Parent = nil
+			v.gui.Enabled = false
+		end
+	end
+
+	self:OnEnabled(enabled)
+end
+
+function Panel:EvaluatePositioning(cameraCF, cameraRenderCF, userHeadCF)
+	if self.panelType == Panel3D.Type.Floor then
+		--Floor panels simply... go on the floor.
+		--Panel will be in camera's local space (which is assumed to be horizontal),
+		--and floorRotation is derived from the user's head rotation (yaw only)
+		local floorCF = cameraCF * CFrame.new(0, -headHeightFromFloor * currentHeadScale, 0) * floorRotation
+		self:SetPartCFrame(floorCF * pointUpCF)
+	elseif self.panelType == Panel3D.Type.Fixed then
+		--Places the panel in the camera's local space, but doesn't follow the user's head.
+		--Useful if you know what you're doing. localCF can be updated in PreUpdate for animation.
+		local cf = self.localCF - self.localCF.p
+		cf = cf + (self.localCF.p * currentHeadScale)
```

Comment:
What is happen with this `cf` variable? Are we using it somewhere I just missed it?